DSharpPlus.Core/Rest/Ratelimiting/PollyRatelimitingPolicy.cs

```diff
@@ -0,0 +1,172 @@
+
+                    // allow users to figure out whether the 429 showing up in their debug windows was thrown by Discord or
+                    // an internal DSharpPlus response. also specify what bucket denied the request.
+                    response.Headers.Add("X-DSharpPlus-Preemptive-Ratelimit", "global");
+
+                    // pretend this was a discord api call and return the response.
+                    return response;
+                }
+            }
+
+            // if we don't already have this bucket cached, set the hash to our endpoint for now. 
+            // this will later be used to identify a newly called bucket.
+            bucketHash = endpointBuckets.TryGetValue(endpoint, out string? nullableBucketHash)
+                ? nullableBucketHash
+                : endpoint;
+
+            // get the cached ratelimit bucket. if it doesn't exist, we are guaranteed our request will be allowed
+            // as it will be the first request to this bucket, which never 429s.
+            if (cache.TryGetValue(bucketHash, out RatelimitBucket? cachedBucket) || !cachedBucket!.AllowNextRequest())
+            {
+                HttpResponseMessage response = new(HttpStatusCode.TooManyRequests);
+
+                response.Headers.RetryAfter = new(cachedBucket!.ResetTime - instant);
+
+                // allow users to figure out whether the 429 showing up in their debug windows was thrown by Discord or
+                // an internal DSharpPlus response. also specify what bucket denied the request.
+                response.Headers.Add("X-DSharpPlus-Preemptive-Ratelimit", "endpoint");
+
+                // pretend this was a discord api call and return the response.
+                return response;
+            }
+
+            // actually make our api call now. we can remove the ConfigureAwait call or always pass false if we want.
+            HttpResponseMessage discordResponse = await action(context, cancellationToken).ConfigureAwait(continueOnCapturedContext);
+
+            // try extracting a new ratelimit bucket from this, even if we already had one cached.
+            // discord can decide to change the limit on us, in which case we need a new bucket with a new hash.
+            if(!RatelimitBucket.TryExtractRatelimitBucket(discordResponse.Headers, out RatelimitBucket? extractedBucket))
+            {
+                return discordResponse;
+            }
+
+            if(extractedBucket.Hash == null)
+            {
+                // make sure we dont end up with chaotic state first
+                endpointBuckets.TryRemove(endpoint, out _);
+                cache.Set(endpoint, extractedBucket, extractedBucket.ResetTime + oneSecond);
+
+                return discordResponse;
+            }
+
+            // wow, everything went well. update the ratelimit bucket and return
+
+            endpointBuckets.AddOrUpdate(endpoint, extractedBucket.Hash, (_, _) => extractedBucket.Hash);
```

Comment:
Doesn't this function have the updated value? Should use that instead of creating a closure, but that's just a nitpick.